### ISSUE TITLE
Refactor unit test layout and naming to be more BDD orientated

### ISSUE THREAD

**Issue by lholman** [lloydstone/femah#52]
I've been pretty verbose with unit test naming, for example.

``` csharp
[Test]
public class FemahApiTests
{
   public void ProcessPutRequestSetsHttpStatusCodeTo200AndReturnsUpdatedFeatureSwitchInResponseBodyIfPutRequestBodyIsValidContainingASingleJsonFeatureSwitchWithAStateChange()
   {
      ...
   }
   [Test]
   public void ProcessPutRequestSetsHttpStatusCodeTo405AndProvidesAccurateErrorMessageInBodyIfPutRequestIsMissingParameterInUrl()
   {
      ...
   }
}
```

Which is all well and good but is not scaling particularly well, with too many tests in a class and names being too long to be of value and in actual fact pushing UI constraints in the likes of TeamCity and build logs.

Moving towards a convention more along the lines of the following may improve discoverability of and reading of unit tests and therefore their value within the project, not to mention make them easier to consume within a build report.

``` csharp
public class WhenProcessingApiPutRequests
{
   [Test] 
   public void GivenAValidRequestThenHttpStatusCode200AndUpdatedFeatureSwitchIsReturned()
   {
      ...
   }
   [Test] 
   public void GivenTheRequestIsMissingAParameterInUrlThenHttpStatusCode405AndErrorMessageIsReturned()
   {
      ...
   }
}
```

**Comment by hotstone** [comment]
Sorry, I'm a bit late to the party here.

So, I've been looking at the new naming convention and it looks great.  I like the way it reads - it's very similar to the convention we used when I first began doing TDD:

```
[TestFixture]
public class FemahTests
{
        [Test]
        public void WhenCallingIsFeatureOn_ThenNoExceptionIsThrown()
        {
        }
}
```

Except obviously we're moving the "When" clause up to be the name of the fixture, and distinguishing variations by the addition of the "And" clause:

```
[TestFixture]
public class WhenCallingIsFeatureOn
{
        [Test]
        public void AndInvokingWithoutInitialising_ThenNoExceptionIsThrown()
        {
        }
}
```

In essence, the difference is that I normally generate a fixture per class under test, while this seems to be a fixture per method of executing the test.

I'm a bit confused though about how we should be handling situations like testing different classes that have a similar method of execution.  Eg. different switch types:

```
    public class RoleBasedFeatureSwitchTests
    {
        [Test]
        public void ReturnsTrueWhenUserInRole()
        {
            var featureSwitch = new RoleBasedFeatureSwitch
            {
                IsEnabled = true,
                Name = "testRoleBasedFeatureSwitch"
            };
            featureSwitch.AcceptedRoles.Add(_testUserRole);

            var result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(true);
        }
```

and

```
    public class PercentageFeatureSwitchTests
    {
        [Test]
        public void UsesValueFromCookieWhenCookieExists()
        {
           var featureSwitch = new PercentageFeatureSwitch()
            {
                IsEnabled = true,
                Name = "testPercentageFeatureSwitch"
            };

            _cookies.Add(new HttpCookie(featureSwitch.Name, true.ToString()));
            var result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(true);

            _cookies.Clear();
            _cookies.Add(new HttpCookie(featureSwitch.Name, false.ToString()));
            result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(false);
        }

        [Test]
        public void SetsCookieWhenNoCookieExists()
        {
            var featureSwitch = new PercentageFeatureSwitch()
            {
                IsEnabled = true,
                Name = "testPercentageFeatureSwitch"
            };

            var result = featureSwitch.IsOn(_femahContext);

            _cookies.Count.ShouldBe(1);
            _cookies[0].Name.ShouldBe(featureSwitch.Name);
            _cookies[0].Value.ShouldBe(result.ToString());
        }
```

(I think we can agree my naming convention isn't the best.)

Would these be aggregated?

```
    public class WhenCallingIsOn
    {
        [Test]
        public void AndIsAPercentageFeatureSwitch_AndACookieExists_ThenUsesValueFromCookie()
        {
           var featureSwitch = new PercentageFeatureSwitch()
            {
                IsEnabled = true,
                Name = "testPercentageFeatureSwitch"
            };

            _cookies.Add(new HttpCookie(featureSwitch.Name, true.ToString()));
            var result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(true);

            _cookies.Clear();
            _cookies.Add(new HttpCookie(featureSwitch.Name, false.ToString()));
            result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(false);
        }

        [Test]
        public void AndIsAPercentageFeatureSwitch_AndACookieExists_ThenSetsCookieWithValue()
        {
            var featureSwitch = new PercentageFeatureSwitch()
            {
                IsEnabled = true,
                Name = "testPercentageFeatureSwitch"
            };

            var result = featureSwitch.IsOn(_femahContext);

            _cookies.Count.ShouldBe(1);
            _cookies[0].Name.ShouldBe(featureSwitch.Name);
            _cookies[0].Value.ShouldBe(result.ToString());
        }

        [Test]
        public void AndIsARoleBasedSwitch_AndUserIsInRole_ThenReturnsTrue()
        {
            var featureSwitch = new RoleBasedFeatureSwitch
            {
                IsEnabled = true,
                Name = "testRoleBasedFeatureSwitch"
            };
            featureSwitch.AcceptedRoles.Add(_testUserRole);

            var result = featureSwitch.IsOn(_femahContext);
            result.ShouldBe(true);
        }
   }
```

Is there some way we could group by the thing being tested AND by the method of testing?  I really like the new naming convention, but don't want to end up with all the tests for all feature switch types in one big fixture.

Maybe `WhenRoleBasedFeatureSwitchCallingIsOn` for the fixture name?  (But better.)

(Sorry, I've had to dash this off quite quickly).

Thoughts guys?

**Comment by lholman** [comment]
It's a valid point @hotstone, apologies, not something myself and Mark spotted when we agreed on the convention.  As you'll see, the main driver was to make the tests far more readable which I think we've achieved, but you're correct, we need to nail this one to make in manageable.

Mark, can we use the parametised tests approach as we discussed here https://github.com/lloydstone/femah/pull/57 to give us that grouping during test execution?

@hotstone Do you want to ensure we maintain a logical grouping within the IDE too?

**Comment by AwaNoodle** [comment]
Regarding grouping in the IDE: VS Test Runner and ReSharper both let you display tests under the class. This lets you have a nice easy to read break down of the When part (so what you're doing) and then see all the sub clause AND parts underneath:

![groupbyclass](https://f.cloud.github.com/assets/3598109/2516536/04b2bf50-b44f-11e3-8191-65291e8f56a7.PNG)

Personally, I think this makes it nice and easy to read the test and to hide away ones I may not be focusing on at present. The current single test class makes this difficult.

**Comment by AwaNoodle** [comment]
Regarding the naming: I am wondering if I am getting the higher level When part correctly on some of these test. I am inferring them based on the accept part of the tests and how they are region grouped at present. Even if the test are moved out of the big test class and those regions into separate classes I still think there is a win for easy of use and readability, it doesn't mean the grouping is logical to the feature or functionality being tested at a macro level.

The examples make me feel that you are testing a different feature, role based features and percentage features which would be different points on a feature list. I'd try to keep them separate. If there were some commonalities between them, like core set up of the feature switch engine, which would require all the tests to change with it, then I'd probably use a factory or fixture object to centralize those bits and keep the tests independent. Tests that follow exactly the same process but only defer with source data are exposing the same feature so I would set those up with param tests. Like some of the tests for return data where the process was the same, only the source and expected out was different. The feature of returning the appropriate status code and error stayed the same throughout.

**Comment by lholman** [comment]
@hotstone @AwaNoodle right, I think we're in a little bit of a pickle.  @hotstone has some valid questions about being able to identify what class is under test, I agree, @AwaNoodle I think you and I have maybe missed that when we concentrated on trying to reduce the verbosity of the test names.
What I suggest is we take all the refactored tests and sanity check them between the 3 of us and come to a common agreement, so to kick things off, the first batch.

Referring to TeamCity build output here: http://bit.ly/1g0llhh

Some thoughts....
- How do we know these tests are testing the API?
- How do we know which class/method is under test, do we explicitly need the name of the class under test, i.e. `ProcessApiRequest` in the name somewhere, I think it's obvious once we know it's for the API and you look at the test?

```
WhenProcessingPutRequests. GivenValidRequest_ThenHttp200AndUpdatedFeatureSwitchBodyIsReturned (Femah.Core.Tests.dll:Femah.Core.Tests)
WhenProcessingPutRequests. GivenRequestWithInvalidEnabledValue_ThenHttpCode400AndErrorAreReturned (Femah.Core.Tests.dll:Femah.Core.Tests)
WhenProcessingPutRequests. GivenRequestMissingParameter_ThenHttp405AndAccurateErrorMessageBodyIsReturned (Femah.Core.Tests.dll:Femah.Core.Tests)
WhenProcessingPutRequests. GivenRequestWithInvalidJson_ThenHttpCode400AndErrorAreReturned (Femah.Core.Tests.dll:Femah.Core.Tests)
WhenProcessingPutRequests. GivenRequestNotForFeatureSwitchService_ThenHttp405AndAccurateErrorMessageBodyIsReturned
WhenProcessingPutRequests. GivenRequestWithInvalidFeatureType_ThenHttpCode400AndErrorAreReturned (Femah.Core.Tests.dll:Femah.Core.Tests)
```

I think if we moved all the above tests in to a `Femah.Api.Tests.dll` assembly we would then know we're testing the API.

**Comment by AwaNoodle** [comment]
I'm thinking that highlighting specific class under test doesn't help you as much as the feature under test. If we were testing System.String we would have a very large single test class working over each method available. To an extent, this is what contains the Femah tests atm. Like all classes, these spin out of control quite quickly. 

Alternatively we can have a class like WhenSplittingAString which highlights and bounds everything relevant to that feature. This class would be smaller and easier to parse for the reader. The tests would be easy to pick out using a test runner than can group by class. From the outside finding what you want seems to me a little easier than looking through one big class or regions.

**Comment by lholman** [comment]
@AwaNoodle @hotstone do you agree with my suggestions in my previous comment then, is that enough visibility @hotstone ?

**Comment by hotstone** [comment]
What was your suggestion sorry?  Moving to a different class library ?  So we'd have Femah.FeatureBasedFeatureSwitch.Tests.dll?

**Comment by lholman** [comment]
@hotstone Read up to my comment yesterday if you can, I gave an example around API tests, I proposed we go through what's been refactored so far to make sure we're all happy.

**Comment by AwaNoodle** [comment]
@lholman I'd worry that the further to the right the identifier is, the less likely someone will be looking to pick up on it. I think the module name would be analogous to the class name in my previous comment. It would define a loose grouping around API tests but you'd still need something more to identify the feature set you're testing for each fixture.

**Comment by lholman** [comment]
@AwaNoodle @hotstone Answers to my question from here please https://github.com/lloydstone/femah/issues/52#issuecomment-38671240
What I have pasted is what we have so far, is that enough? We're stalling on this a bit.

**Comment by hotstone** [comment]
@lholman 

> How do we know these tests are testing the API?

We don't, unless we move them into the API.Tests namespace as you mentioned.

> How do we know which class/method is under test, do we explicitly need the name of the class under test, i.e. ProcessApiRequest in the name somewhere, I think it's obvious once we know it's for the API and you look at the test?

We don't know which class is under tests, although we do know which method is under test (it's in the test fixture name); I would say we do explicitly need the class name in the name somewhere.

Just to clarify what my original question was: How do we deal with different classes under test that have the same method of invocation, if we're grouping tests by method of invocation?

**Comment by lholman** [comment]
@hotstone Cool, so agreed on moving them to a different namespace.

> Just to clarify what my original question was: How do we deal with different classes under test that have the same method of invocation, if we're grouping tests by method of invocation?

Have you some examples of how we could change the `WhenProcessingPutRequests` ones to accommodate that?

**Comment by AwaNoodle** [comment]
Can we have the tests divided by:
1. Assembly name, i.e. Femah.Api.Tests
2. Feature under test, i.e. WhenSettingFeatureSwitch
3. Test for the feature, i.e. AndFeatureExists_ThenFeatureIsSetToEnabled

**Comment by lholman** [comment]
@AwaNoodle You seem to be mixing your example names there, we wouldn't have FeatureSwitch tests in Femah.Api.Tests, would we?

**Comment by hotstone** [comment]
@Awanoodle Would you mind providing an example using the RoleBasedFeatureSwitch and PercentageBasedFeatureSwitch examples I used earlier?  I think I'm just getting confused on how we differentiate between different classes (RoleBasedFeatureSwitch and PercentageBasedFeatureSwitch) that have the same feature to test (IsOn).

**Comment by hotstone** [comment]
@AwaNoodle  Just to follow up, earlier you wrote: 

> The examples make me feel that you are testing a different feature, role based features and percentage features which would be different points on a feature list. I'd try to keep them separate.

Which sounds good to me, but I'm not sure how you're suggesting to keep them separate.  Up till now I've done that by placing them in separate test fixtures (eg. "RoleBasedFeatureSwitchTests.cs" and "PercentageBasedFeatureSwitchTests.cs").  

Maybe a naming system like WhenCallingIsFeatureOnInARoleBasedFeatureSwitch.cs?

**Comment by hotstone** [comment]
Looking at the discussion, it seems to me that the fundamental problem here is that we're skipping the 'given' in the usual given-when-then tuplet.  Since the 'given' establishes the context within which the action (in the 'when') takes place, it makes sense to me that I'm getting confused about _what_  the action is being done to.

Eg. if it were:

```
GivenARoleBasedFeatureSwitch.WhenCallingIsOn_AndTheUserIsInTheRole_ThenFalse()
GivenAPercentageBasedFeatureSwitch.WhenCallingIsOn_AndTheNumberIsBeneathTheThreshold_ThenTrue()
```

then I'd understand how we differentiate similar tests against different classes.

Without the 'given' parts, I get confused:

```
WhenCallingIsOn.AndTheUserIsInTheRole_ThenFalse()
WhenCallingIsOn.AndTheNumberIsBeneathTheThreshold_ThenTrue()
```

**Comment by lholman** [comment]
@hotstone I like that, but what would the namespace and class name be for that example, something like this?

```
namespace Femah.Core.Tests
{
    public class FeatureSwitchTests
    {
        [Test]
        public void GivenARoleBasedFeatureSwitch.WhenCallingIsOn_AndTheUserIsInTheRole_ThenFalse()
        {
```

My concern with that is it potentially takes us right back to the initial problem https://github.com/lloydstone/femah/issues/52#issue-29741155 which was length of test name? 

On another note, in specflow, you would have different test definitions for the Given, When, Then tuplets, are we getting muddled with unit tests versus more integration/acceptance tests?

**Comment by lholman** [comment]
@AwaNoodle @hotstone I don't think this discussion is really getting us anywhere.  It seems that the decision Mark and I made to use the Given/When/Then syntax, primarily to try and make some existing verbosely named tests more readable has resulted in a convention that doesn't allow us to easily differentiate similar tests against different classes, as Matt has identified.  I don't want to throw this work away and am committed to getting the right solution.

We now have a few options.
1. Continue on with the Given/When/Then syntax refactoring - Probably not the right approach seeing as we can't come to an agreement.
2. Review the ApiTests that I have written, maybe there's something fundamentally wrong there?
3. Revert back to where we were, identifying the class under test in the test class name? - Again probably not that straight forward, at least without looking at number 2. above as names were getting very long.
4. Think of a different approach.  I've had a look around and the approach used by the guys on the github api c# library https://github.com/octokit/octokit.net/blob/master/Octokit.Tests/GitHubClientTests.cs where they are subclassing the methods/properties under test seems really clean.

I vote for 2 and 4.

Please let's decide and move on before we waste anyone more time and effort

**Comment by lholman** [comment]
A couple of examples using the [Octokit.Net](https://github.com/octokit/octokit.net/blob/master/Octokit.Tests/GitHubClientTests.cs) approach of identifying both the class and method/property under test in class and sub-class names. I'm not sure if this is a common approach, or overkill, it seems to work nicely though.

Instead of `GivenRequestContainsInvalidJson_ThenHttp400AndGenericErrorMessageBodyIsReturned` from  https://github.com/lloydstone/femah/blob/master/Femah.Core.Tests/WhenProcessingPutRequests.cs 

``` C#
namespace Femah.Core.Tests
{
    public class ProcessApiRequestTests
    {
        public class TheProcessPutRequestMethod
        {
            [Test]
            public void ReturnsHttp400AndGenericErrorMessageBody_WhenRequestContainsInvalidJson
        {
```

Instead of `ThenResponseHasCorrectEncodingAndContentType` from  https://github.com/lloydstone/femah/blob/master/Femah.Core.Tests/WhenAccessingTheService.cs

``` C#
namespace Femah.Core.Tests
{
    public class FemahApiHttpHandlerTests
    {
        public class TheProcessRequestMethod
        {
            [Test]
            public void EnsuresHttpReponseHasCorrectEncodingAndContentType
        {
```

Instead of `AndInvokingWithoutInitialising_ThenNoExceptionIsThrown` from  https://github.com/lloydstone/femah/blob/master/Femah.Core.Tests/WhenCallingIsFeatureOn.cs

``` C#
namespace Femah.Core.Tests
{
    public class FemahTests
    {
        public class TheIsFeatureOnMethod
        {
            [Test]
            public void DoesNotThrowAnException_WhenInvokingWithoutInitialising
        {
```

Instead of `UsesValueFromCookieWhenCookieExists` from  https://github.com/lloydstone/femah/blob/master/Femah.Core.Tests/PercentageFeatureSwitchTests.cs

``` C#
namespace Femah.Core.Tests
{
    public class PrecentageFeatureSwitchTests
    {
        public class TheIsOnMethod
        {
            [Test]
            public void UsesValueFromCookieWhenCookieExists
        {
```

I'm not sure what these would look like in the NUnit TestResult.xml and subsequently in the ReSharper UI and TeamCity build output, happy to give it a go though if we see value it it?

**Comment by hotstone** [comment]
@lholman This (nested classes) is pretty much what I was thinking too, actually - I was looking at how other libraries name their unit tests and this is what I found to be quite common.  And it makes sense to me  - it means the name of the SUT is both in the filenames and in test fixture name

So I vote we plow ahead with that!

**Comment by AwaNoodle** [comment]
Nested classes does lead to very large classes that become hard to navigate. Clean code and all that :)

Could we extend the namespace like:

``` C#
namespace Femah.Core.Tests.PrecentageFeatureSwitchTests
{
     public class TheIsOnMethod
     {
            [Test]
            public void UsesValueFromCookieWhenCookieExists
            {
                   // Awesome goes here
            }
```

and divide tests into solution sub folders with the SUT name?

**Comment by lholman** [comment]
@AwaNoodle Good point, a subtle variation, using namespaces instead of classes for the outermost layer.  That said I'd like to give it a go with subclasses and refactor out if and when we need to, it seems to be a popular choice with sub-classes too http://haacked.com/archive/2012/01/02/structuring-unit-tests.aspx/.

The main reason I'd like to go with sub-classes is so that we maintain a consistency with the name of the top-level class matching that of the resultant .cs file, in my examples that would be `PrecentageFeatureSwitchTests.cs` with the assembly `Femah.Core.Tests.dll`, with a namespace approach we would not have one top-level class, causing confusion with file naming, which I believe in turn would make the structure messy.
I'm sure both will give us the visibility here but femah is a small library and I don't think we're going suffer much mess with sub-classes in the foreseeable future.

**Comment by hotstone** [comment]
@lholman @AwaNoodle Yep, I agree with this - having the name of the SUT in the outermost class means it's also in the filename.  

Let's go with this and if we do run into any issues, we can reconsider then.

**Comment by lholman** [comment]
@AwaNoodle @hotstone Ok, now we're agreed lets get going again, it's taken nearly a week to get here.  @AwaNoodle are you keen to continue helping with this, if so I suggest we get the changes you've already made in to this format and then progress on with the others.  I'm happy to help too.

**Comment by AwaNoodle** [comment]
@lholman  I'm fired up to get on with this :) Just to be sure, what we are doing is:
- Assembly stays as is
- Filename is the SUT
- Class name is the SUT
- Subclasses are used for the When.... or similar naming
- Methods contain the If_Then or similar naming

**Comment by AwaNoodle** [comment]
I'll start with re-factoring the parts I've already worked on. Anyone up for taking on bits that are in the main test class? So:
- [ ] Refactor classes in When... format to follow above style
- [ ] Move existing tests out of main test class to follow above style

**Comment by lholman** [comment]
@AwaNoodle Re: https://github.com/lloydstone/femah/issues/52#issuecomment-39213047 I believe it should be 
1. Assembly stays as is
2. Filename is the SUT
3. Class name is the SUT and matches the filename
4. Subclasses are used to denote the method or property under test, i.e. 'The **IsOn** Method'
5. Methods contain an appropriate test name that describes the assertion.

**Comment by AwaNoodle** [comment]
@lholman Ok, I'll get on with it :)
